ttnn/cpp/ttnn/operations/embedding_backward/device/embedding_backward_program_factory.cpp

```diff
@@ -0,0 +1,182 @@
+
+    // reader
+
+    bool index_stick_size_is_power_of_two = is_power_of_two_at_least_32(index_page_size);
+    uint32_t index_log2_stick_size = index_stick_size_is_power_of_two ? log2(index_page_size) : 0;
+
+    std::vector<uint32_t> reader_compile_time_args = {
+        grad_is_dram,
+        index_is_dram,
+        out_is_dram,
+        index_page_size,
+        index_stick_size_is_power_of_two,
+        index_log2_stick_size,
+        index_tensor.get_dtype() == DataType::BFLOAT16,
+        max_tiles_per_core,
+        batch_size,
+        seq_len_tiles,
+        num_embeddings_tiles};
+
+    auto reader_kernel_id = tt_metal::CreateKernel(
+        program,
+        "ttnn/cpp/ttnn/operations/embedding_backward/device/kernels/dataflow/reader_embedding_backward.cpp",
+        all_cores,
+        tt_metal::ReaderDataMovementConfig(reader_compile_time_args));
+
+    std::vector<uint32_t> reader_runtime_args = {
+        grad_tensor_buffer->address(),
+        index_tensor_buffer->address(),
+        out_buffer->address(),
+        embedding_tiles,  // how many pages to skip to get to the next row
+        0,                // offset to the first tile in a row
+        0,                // how many tiles to process in a row
+    };
+
+    auto compute_kernel_id = tt_metal::CreateKernel(
+        program,
+        "ttnn/cpp/ttnn/operations/embedding_backward/device/kernels/compute/embedding_backward.cpp",
+        all_cores,
+        tt_metal::ComputeConfig{.compile_args = {max_tiles_per_core, input_height_tiles}});
+
+    ////////////////////////////////////////////////////////////////////////////
+    //                 Run-time arguments
+    ////////////////////////////////////////////////////////////////////////////
+
+    auto cores = corerange_to_cores(all_cores);
+    uint32_t offset = 0;
+    for (auto core : cores) {
+        reader_runtime_args[4] = offset;
+        if (core_group_1.core_coord_in_core_ranges(core)) {
+            reader_runtime_args[5] = num_tiles_per_core_group_1;
+        } else {
+            reader_runtime_args[5] = num_tiles_per_core_group_2;
+        }
+        SetRuntimeArgs(program, reader_kernel_id, core, reader_runtime_args);
+        SetRuntimeArgs(program, compute_kernel_id, core, {reader_runtime_args[5]});
+
+        offset += reader_runtime_args[5];
+    }
+
+    auto override_runtime_args_callback = [reader_kernel_id, cores, device](
+                                              const Program &program,
+                                              const std::vector<Buffer *> &input_buffers,
+                                              const std::vector<Buffer *> &output_buffers) {
+        auto grad_dram_buffer = input_buffers.at(0);
+        auto index_dram_buffer = input_buffers.at(1);
+        auto output_dram_buffer = output_buffers.at(0);
+
+        for (const auto &core : cores) {
+            {
+                auto &runtime_args = GetRuntimeArgs(program, reader_kernel_id, core);
```

Comment:
Change to this for better perf?
```suggestion
        auto &runtime_args_by_core = GetRuntimeArgs(program, reader_kernel_id);
        for (const auto &core : cores) {
            {
                auto &runtime_args = runtime_args_by_core[core.x][core.y];
```